sklearn/utils/tests/test_utils.py

```diff
@@ -262,6 +265,61 @@ def check_warning(*args, **kw):
         assert type(actual) is type(expected)
 
 
+@pytest.mark.parametrize(
+    ['source', 'message', 'is_long'],
+    [
+        ('ABC', string.ascii_lowercase, False),
+        ('ABCDEF', string.ascii_lowercase, False),
+        ('ABC', string.ascii_lowercase * 3, True),
+        ('ABC' * 10, string.ascii_lowercase, True),
+        ('ABC', string.ascii_lowercase + u'\u1048', False),
+    ])
+@pytest.mark.parametrize(
+    ['time', 'time_str'],
+    [
+        (0.2, '   0.2s'),
+        (20, '  20.0s'),
+        (2000, '33.3min'),
+        (20000, '333.3min'),
+    ])
+def test_message_with_time(source, message, is_long, time, time_str):
+    out = message_with_time(source, message, time)
+    if is_long:
+        assert len(out) > 70
+    else:
+        assert len(out) == 70
+
+    assert out.startswith('[' + source + '] ')
+    out = out[len(source) + 3:]
+
+    assert out.endswith(time_str)
+    out = out[:-len(time_str)]
+    assert out.endswith(', elapsed=')
+    out = out[:-len(', elapsed=')]
+    assert out.endswith(message)
+    out = out[:-len(message)]
+    assert out.endswith(' ')
+    out = out[:-1]
+
+    if is_long:
+        assert not out
+    else:
+        assert list(set(out)) == ['.']
+
+
+@pytest.mark.parametrize(
+    ['message', 'expected'],
+    [
+        ('hello', message_with_time('ABC', 'hello', 0.1) + '\n'),
+        ('', message_with_time('ABC', '', 0.1) + '\n'),
+        (None, ''),
+    ])
+def test_log_elapsed(message, expected, capsys):
+    with log_elapsed('ABC', message):
+        time.sleep(0.1)
+    assert capsys.readouterr()[0] == expected
```

Comment:
`time.sleep` is not deterministic https://stackoverflow.com/questions/1133857/how-accurate-is-pythons-time-sleep

This could be good enough, but it's probably safer to use the `monkeypatch` pytest fixture to make, `timeit.timer()` return pre-defined values before and inside the context manager to avoid random failures on exotic systems.